### ISSUE TITLE
Empty batch requests being sent to server

### ISSUE THREAD

**Issue by sitaram-kalluri** [atsign-foundation/at_client_sdk#731]
**Describe the bug**
When uncommitted entries contains the keys are not synced to secondary server (the cached keys and keys that are local to client), 
the sync service skips the entries. But this causes the delay keys to sync because the non sync keys in uncommitted entries list is being iterated and an empty batch requests are being sent to server.

Client Logs:

```
  FINER|2022-09-29 07:20:25.629525|SyncService|cached:@ce2e2:phonebeef3ec1-7696-4f11-ac7a-76851125cca6.e2etest@ce2e1 is skipped. cached keys will not be synced to cloud secondary 
  
  FINER|2022-09-29 07:20:25.629552|SyncService|cached:@ce2e2:phone55df8976-f61f-4fb3-b547-56bbcfae9aaf.e2etest@ce2e1 is skipped. cached keys will not be synced to cloud secondary 
  
  FINER|2022-09-29 07:20:25.629567|SyncService|cached:@ce2e2:phone094f55ce-cd83-4218-bb6a-43e38cc53471.e2etest@ce2e1 is skipped. cached keys will not be synced to cloud secondary 
  
  FINER|2022-09-29 07:20:25.629576|SyncService|cached:@ce2e2:phone540142bc-a9ce-4c7a-aa91-e3d3babd7304.e2etest@ce2e1 is skipped. cached keys will not be synced to cloud secondary 
  
  FINER|2022-09-29 07:20:25.629589|SyncService|cached:@ce2e2:phonea20590dd-c87c-43a3-8e65-253c4b4cc66b.e2etest@ce2e1 is skipped. cached keys will not be synced to cloud secondary 
  
  FINER|2022-09-29 07:20:25.629636|AtLookup|SENDING: batch:[]
   
  
  FINER|2022-09-29 07:20:25.708053|OutboundMessageListener|RECEIVED data:[] 
  
  FINER|2022-09-29 07:20:25.717212|SyncService|batch result:data:[] 
```

Server Logs:

```
ce2e2_secondary.1.vtcdxu0eh32m@cicd2x64    | INFO|2022-09-29 07:20:26.287185|InboundListener|RCVD: [_431fb7f9-5fe5-4fa6-9d67-b0e3ebc1574b] batch:[] 
ce2e2_secondary.1.vtcdxu0eh32m@cicd2x64    | 
ce2e2_secondary.1.vtcdxu0eh32m@cicd2x64    | INFO|2022-09-29 07:20:26.287396|InboundConnectionImpl|SENT: [_431fb7f9-5fe5-4fa6-9d67-b0e3ebc1574b] data:[]
```

<details>
<summary> Complete logs describing the delay in sync process</summary>

```
FINER|2022-09-29 07:20:25.629180|SyncService|syncing to remote. Total uncommitted entries: 423 
  
  FINER|2022-09-29 07:20:25.629525|SyncService|cached:@ce2e2:phonebeef3ec1-7696-4f11-ac7a-76851125cca6.e2etest@ce2e1 is skipped. cached keys will not be synced to cloud secondary 
  
  FINER|2022-09-29 07:20:25.629552|SyncService|cached:@ce2e2:phone55df8976-f61f-4fb3-b547-56bbcfae9aaf.e2etest@ce2e1 is skipped. cached keys will not be synced to cloud secondary 
  
  FINER|2022-09-29 07:20:25.629567|SyncService|cached:@ce2e2:phone094f55ce-cd83-4218-bb6a-43e38cc53471.e2etest@ce2e1 is skipped. cached keys will not be synced to cloud secondary 
  
  FINER|2022-09-29 07:20:25.629576|SyncService|cached:@ce2e2:phone540142bc-a9ce-4c7a-aa91-e3d3babd7304.e2etest@ce2e1 is skipped. cached keys will not be synced to cloud secondary 
  
  FINER|2022-09-29 07:20:25.629589|SyncService|cached:@ce2e2:phonea20590dd-c87c-43a3-8e65-253c4b4cc66b.e2etest@ce2e1 is skipped. cached keys will not be synced to cloud secondary 
  
  FINER|2022-09-29 07:20:25.629636|AtLookup|SENDING: batch:[]
   
  
  FINER|2022-09-29 07:20:25.708053|OutboundMessageListener|RECEIVED data:[] 
  
  FINER|2022-09-29 07:20:25.717212|SyncService|batch result:data:[] 
  
  FINER|2022-09-29 07:20:25.717265|SyncService|cached:@ce2e2:phoned3643ecb-897b-427e-b973-a61e307a5c34.e2etest@ce2e1 is skipped. cached keys will not be synced to cloud secondary 
  
  FINER|2022-09-29 07:20:25.717284|SyncService|cached:@ce2e2:phonea04e879a-ceb8-4dbf-a0ad-70612ba37a7b.e2etest@ce2e1 is skipped. cached keys will not be synced to cloud secondary 
  
    
  FINER|2022-09-29 07:20:28.357460|SyncService|batchId:1 key:_latestnotificationidv2.keystream@ce2e2 
  
  FINER|2022-09-29 07:20:28.357674|SyncService|batchId:2 key:public:publickey@ce2e2 
  
  FINER|2022-09-29 07:20:28.357849|SyncService|batchId:3 key:_latestnotificationidv2.wavi@ce2e2 
  
  FINER|2022-09-29 07:20:28.358006|SyncService|batchId:4 key:_latestnotificationidv2.wavi@ce2e2 
  
  FINER|2022-09-29 07:20:28.358163|SyncService|batchId:5 key:_latestnotificationidv2.wavi@ce2e2 
  
  FINER|2022-09-29 07:20:28.358238|AtLookup|SENDING: batch:[{"id":1,"command":"update:ttl:0:ttb:0:ccd:false:isBinary:false:isEncrypted:false:_latestnotificationidv2.keystream@ce2e2 mv0p9rEK2+rgQ8RG1/qiofXjqxFyLZItQUicQWvWjqvmSEe1Pl35QqWgdSwlLyd97KXnLJrvhd00vVuJNvmWlu0br8wcmBvnYerJdd5RHSXiZKMn1jUTLSVyoRHB+qsSs35kcYKJV00OAlAZQXiyusJQ5XgV1hC+t7ZKSYyfiIDNo423y64LqZUpwh1TZ+tI0qA0JBfwKKvXSsMTPMVsViHyG3YDE7pi8j7LRs4PSVroqilJumYEZnfLHw/epMcpOov7r877HO+wBER/3C9312xyd/XpjSYTM+7yP8AJoeODYkgga5dbnI9KcSIXlQs1v2+jPRUVahQ9W2DQOwmpUkWAe9o0q272xBqNDwcVc1qC3NcMhjLTxdB7wtP+ENiiXG615IuNfCh/kQ+87wrzTQ=="***{"id":2,"command":"update:ttl:0:ttb:0:ccd:false:isBinary:false:isEncrypted:false:public:publickey@ce2e2 MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAls/anFf/q9OdIIGICEwEosKHVgDe8GiVMLWmdAbVVYdORAcUsWedNYJ5fuWBtiToaQE/NXB4e8J3CgpbQVeoOgwdKM7DTl7mS0JT10bgxVaKf01EYE9+4I2/sVpRKe/rvQ3kr2y5k5mPIxm/0GkMRidYRTriEu1bEky152RLEjLNXTLBnI/kvWJDApkbD9FwMyBFZBRHjb2kg/jV+eDpv+IFl+uLO/tdpIc+hpa/PsSP6uL9KSJKL03p0BqYQik3oAbdEdN2eyU/mYanmVeRC/YDnGo4OtGjtBni9N4qHavQKE9aMMU2gB5K0JuxRZwdmH8U8TdsR3oeDjMAHT3oGwIDAQAB"***{"id":3,"command":"update:isBinary:false:isEncrypted:false:_latestnotificationidv2.wavi@ce2e2 mv0p9rEK2+q1FpMQhqryofDhphdyLcQkR0jEQTmCjqDuTRW2P1qkHv75JSwlLyd97KXnLJrvhd00vVuONamfkLgcpMwWn0m3YerLdd5RHSflYqNx3jhCLiJxrRLLo/EU6CxhetWIUk1dUFhEQXiyv5ZQ5XxF1hDv5LVPGIrD0YbM89C3y64LqZUpwh1TZ+tI0qA0JBfwKKvXSsMTPMVsViHyG3YDE7pi8j7LRs4PSVroqilJumYEZnfLHw/epMcpOov7r877HOq1D0l+2CF312xyd/XpjSYTM+7yP8AJoeODYkgga5dbnI9KcSIXlQs1v2+jPRUVahQ9W2DQOwmpUkWAe9o0q272xBqNDwcVc1qC3NcMhjLTxdB7wtP+ENiiXG615IuNfCh/kQ+87wrzTQ=="***{"id":4,"command":"update:isBinary:false:isEncrypted:false:_latestnotificationidv2.wavi@ce2e2 mv0p9rEK2+q1FpMQhqryofDhphdyLcQkR0jEQTmCjqDuTRW2P1qkHv75JSwlLyd97KXnLJrvhd00vVuONamfkLgcpMwWn0m3YerLdd5RHSflYqNx3jhCLiJxrRLLo/EU6CxhetWIUk1dUFhEQXiyv5ZQ5XxF1hDv5LVPGIrD0YbM89C3y64LqZUpwh1TZ+tI0qA0JBfwKKvXSsMTPMVsViHyG3YDE7pi8j7LRs4PSVroqilJumYEZnfLHw/epMcpOov7r877HOq1D0l+2CF312xyd/XpjSYTM+7yP8AJoeODYkgga5dbnI9KcSIXlQs1v2+jPRUVahQ9W2DQOwmpUkWAe9o0q272xBqNDwcVc1qC3NcMhjLTxdB7wtP+ENiiXG615IuNfCh/kQ+87wrzTQ=="***{"id":5,"command":"update:isBinary:false:isEncrypted:false:_latestnotificationidv2.wavi@ce2e2 mv0p9rEK2+q1FpMQhqryofDhphdyLcQkR0jEQTmCjqDuTRW2P1qkHv75JSwlLyd97KXnLJrvhd00vVuONamfkLgcpMwWn0m3YerLdd5RHSflYqNx3jhCLiJxrRLLo/EU6CxhetWIUk1dUFhEQXiyv5ZQ5XxF1hDv5LVPGIrD0YbM89C3y64LqZUpwh1TZ+tI0qA0JBfwKKvXSsMTPMVsViHyG3YDE7pi8j7LRs4PSVroqilJumYEZnfLHw/epMcpOov7r877HOq1D0l+2CF312xyd/XpjSYTM+7yP8AJoeODYkgga5dbnI9KcSIXlQs1v2+jPRUVahQ9W2DQOwmpUkWAe9o0q272xBqNDwcVc1qC3NcMhjLTxdB7wtP+ENiiXG615IuNfCh/kQ+87wrzTQ=="***]
   
  
  FINEST|2022-09-29 07:20:28.367926|Monitor|_messageHandler received data 
  
  FINER|2022-09-29 07:20:28.367973|Monitor|received response on monitor: notification: {"id":"43cddf1e-bb9a-4c56-a20c-209a268b858a","from":"@ce2e1","to":"@ce2e2","key":"@ce2e2:phonenumber02d85ecf-ba87-4945-860e-844ab7b9a93a.wavi@ce2e1","value":null,"operation":"delete","epochMillis":1664436028280,"messageType":"MessageType.key","isEncrypted":false*** 
  

```
</details>

Github action run : https://github.com/atsign-foundation/at_client_sdk/actions/runs/3149341230/jobs/5120863575

**Comment by gkc** [comment]
Setting to High priority as it is a bug.

**Comment by sitaram-kalluri** [comment]
Spent 2 SP in PR-47 for root cause analysis

When fetching the uncommitted entries on the client side following happens:

Step 1: Lets say server is at commit Id - 10
Step 2: Lets say the client is at commit id 5, and there are 3 uncommitted entries which are at hive seq 6,7,8 respectively.
Step 3: When sync completes, first the 5 commits in server are pulled to local, then 3 uncommitted entries are sync'ed server. So the lastSyncedEntry will be key with hive seq number 8.
Step 4: Next time, when uncommitted entries are to fetched, all the keys which are higher then hive seq 8 are fetched causing the committed entries to be fetched again.

If the entries contains cached keys, since cached does not get sync'ed to server we skip then on batch requests causing the empty batch requests to be sent to server.

An approach to fix the issue is: Fetch only commit's with null commit id's into uncommitted entries. This approach still have to tested.

**Comment by srieteja** [comment]
Had a productive discussion with Sitaram. We have a possible solution based on an assumption; which we are testing out in-depth and I will be linking the PR after the testing.

**Comment by srieteja** [comment]
The approach used in the linked PR is: when 'enableCommitId' is true (indicating that the persistence server belongs to the client) then a check is placed to ensure only keys without commit Ids are returned.

**Comment by srieteja** [comment]
Moving to PR49 for testing.

**Comment by sitaram-kalluri** [comment]
The changes are in at_persistence_secondary_server and are merged to the trunk. Re-opening the git issue to publish the changes.

**Comment by sitaram-kalluri** [comment]
Pending work is to publish at_persistence_secondary_server package. Hence re-estimating to 0SP

**Comment by sitaram-kalluri** [comment]
The changes are published in at_persistence_secondary_server: v3.0.43.

**Comment by sitaram-kalluri** [comment]
The at_persistence_secondary_server 3.0.43 version is consumed by at_client in v3.0.45 which fixes the issue.
